tools/codegen/dest/lazy_ir.py

```diff
@@ -0,0 +1,387 @@
+            return NamedCType(binds, BaseCType(valueT))
+        elif typ.name == BaseTy.Scalar:
+            return NamedCType(binds, BaseCType(scalarT))
+        else:
+            raise AssertionError(f"TODO add support for type {repr(typ)}")
+    elif isinstance(typ, OptionalType):
+        if str(typ.elem) == 'Tensor':
+            return NamedCType(binds, OptionalCType(BaseCType(valueT)))
+        elif str(typ.elem) == 'ScalarType':
+            return NamedCType(binds, OptionalCType(BaseCType(scalarTypeT)))
+        else:
+            raise AssertionError(f"TODO add support for type {repr(typ)}")
+    elif isinstance(typ, ListType):
+        if str(typ.elem) == 'Tensor':
+            return NamedCType(binds, BaseCType(valueListT))
+        elif str(typ.elem) == 'Tensor?':
+            return NamedCType(binds, ListCType(OptionalCType(BaseCType(valueT))))
+        elif str(typ.elem) == 'int':
+            return NamedCType(binds, BaseCType(intArrayT))
+        else:
+            raise AssertionError(f"TODO add support for type {repr(typ)}")
+    else:
+        raise AssertionError(f"unrecognized type {repr(typ)}")
+
+
+def isValueType(typ: NamedCType) -> bool:
+    """
+    Given a type, determine if it is a Value-like type.  This is equivalent to
+    being Tensor-like, but assumes the type has already been transformed.
+    """
+    if isinstance(typ.type, BaseCType):
+        return typ.type.type.ns == valueT.ns and typ.type.type.name == valueT.name
+    elif isinstance(typ.type, OptionalCType):
+        return isValueType(typ.type.elem)
+    elif isinstance(typ.type, ListCType):
+        return isValueType(typ.type.elem)
+    else:
+        return False
+
+
+def update_schema_for_lazy_ir(func: FunctionSchema) -> FunctionSchema:
+    """
+    We handle Tensor arguments specially as 'IR Values'.
+
+    This function mainly deals with the lists of types buried in named fields of `Arguments`,
+    delegating the actual type conversion to `process_ir_types`
+    """
+    new_args_fields = {}
+    for arg_field in ["pre_self_positional",
+                      "self_arg",
+                      "post_self_positional",
+                      "pre_tensor_options_kwarg_only",
+                      "tensor_options",
+                      "post_tensor_options_kwarg_only",
+                      "out"]:
+        if arg_field == "self_arg" and getattr(func.arguments, "self_arg") is not None:
+            arg = getattr(func.arguments, "self_arg").argument
+            new_args_fields[arg_field] = SelfArgument(Argument(arg.name, process_ir_type(
+                arg.type, arg.is_write, arg.name), arg.default, arg.annotation))
+        elif getattr(func.arguments, arg_field) is not None:
+            new_args_fields[arg_field] = [
+                Argument(
+                    arg.name,
+                    process_ir_type(arg.type, arg.is_write, arg.name), arg.default, arg.annotation) for arg in getattr(func.arguments, arg_field)]
```

Comment:
Hmm... transforming the `FunctionSchema` feels a little sketchy. I also see that `process_ir_type` returns a `NamedCType`, but the constructor for `Argument()` only takes in normal `Type`s?

It looks like the main purpose of this function is to substitute `Tensor` with `ir::Value` in a bunch of places. Instead of transforming the `FunctionSchema`, why can't you do that translation when you're generating each C++ string?

Reply:
I had messier code before, and i refactored to manipulate FunctionSchema since it seemed like a nice way to modularize some of the transformations i wanted to do.

is your main objection that I abuse an existing class that has some semantic meaning, and if I make a new class for Lazy* it'll be cleaner?

Or do you also object to the modification at `FunctionSchema` level vs `Signature` level?  

I do more than s/Tensor-like/Value-like.  I also do stuff to make reference types into concrete types I can store in a class.

Also, I use the resulting modified schema in more places than just generating a decl or defn for a function.  For example I use the resulting modified argument types in places for IR node construction, and, I find it useful to have access to the modified arguments from the perspective of pre_self_positional, vs kwargs, etc - see how I use that in ts_lowering.py